atom/browser/api/atom_api_power_save_blocker.cc

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2015 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "atom/browser/api/atom_api_power_save_blocker.h"
+
+#include "content/public/browser/power_save_blocker.h"
+#include "native_mate/constructor.h"
+#include "native_mate/dictionary.h"
+
+#include "atom/common/node_includes.h"
+
+namespace {
+
+const char kPowerSaveBlockerDescription[] = "Electron";
+
+}  // namespace
+
+namespace mate {
+
+template<>
+struct Converter<content::PowerSaveBlocker::PowerSaveBlockerType> {
+  static bool FromV8(v8::Isolate* isolate,
+                     v8::Local<v8::Value> val,
+                     content::PowerSaveBlocker::PowerSaveBlockerType* out) {
+    using content::PowerSaveBlocker;
+    int type;
+    if (!ConvertFromV8(isolate, val, &type))
+      return false;
+    switch (static_cast<PowerSaveBlocker::PowerSaveBlockerType>(type)) {
+      case PowerSaveBlocker::kPowerSaveBlockPreventAppSuspension:
+        *out = PowerSaveBlocker::kPowerSaveBlockPreventAppSuspension;
+        break;
+      case PowerSaveBlocker::kPowerSaveBlockPreventDisplaySleep:
+        *out = PowerSaveBlocker::kPowerSaveBlockPreventDisplaySleep;
+        break;
+      default:
+        return false;
+    }
+    return true;
+  }
+};
+
+}  // namespace mate
+
+namespace atom {
+
+namespace api {
+
+PowerSaveBlocker::PowerSaveBlocker() {
+}
+
+PowerSaveBlocker::~PowerSaveBlocker() {
+}
+
+void PowerSaveBlocker::UpdatePowerSaveBlocker() {
+  if (power_save_blocker_types_.empty()) {
+    power_save_blocker_.reset();
+    return;
+  }
+
+  // |kPowerSaveBlockPreventAppSuspension| keeps system active, but allows
+  // screen to be turned off.
+  // |kPowerSaveBlockPreventDisplaySleep| keeps system and screen active, has a
+  // higher precedence level than |kPowerSaveBlockPreventAppSuspension|.
+  //
+  // Only the highest-precedence blocker type takes effect.
+  content::PowerSaveBlocker::PowerSaveBlockerType new_blocker_type =
+      content::PowerSaveBlocker::kPowerSaveBlockPreventAppSuspension;
+  for (const auto& element : power_save_blocker_types_) {
+    if (element.second ==
+        content::PowerSaveBlocker::kPowerSaveBlockPreventDisplaySleep) {
+      new_blocker_type =
+          content::PowerSaveBlocker::kPowerSaveBlockPreventDisplaySleep;
+      break;
+    }
+  }
+
+  if (!power_save_blocker_ || new_blocker_type != current_blocker_type_) {
+    scoped_ptr<content::PowerSaveBlocker> new_blocker =
+        content::PowerSaveBlocker::Create(
+            new_blocker_type,
+            content::PowerSaveBlocker::kReasonOther,
+            kPowerSaveBlockerDescription);
+    power_save_blocker_.swap(new_blocker);
```

Comment:
We should probably use `reset` to directly replace original PowerSaveBlocker, PowerSaveBlocker is an RAII-style class so PowerSaveBlocker instances have to be deallocated in the reverse order of how they are allocated, otherwise bad things may happen.